File: front/src/lib/schemas/reference_sounds.ts
"use client";

import { z } from "zod";

import { ClipSchema } from "./clips";
import { TagSchema } from "./tags";

// Reference sound source enum
export const ReferenceSoundSourceSchema = z.enum([
  "xeno_canto",
  "custom_upload",
  "dataset_clip",
]);

export type ReferenceSoundSource = z.infer<typeof ReferenceSoundSourceSchema>;

// Main ReferenceSound schema
export const ReferenceSoundSchema = z.object({
  uuid: z.string().uuid(),
  name: z.string(),
  description: z.string().nullable().optional(),
  notes: z.string().nullable().optional(),
  ml_project_id: z.number().int().optional(),
  ml_project_uuid: z.string().uuid().optional(),
  source: ReferenceSoundSourceSchema,
  xeno_canto_id: z.string().nullable(),
  xeno_canto_url: z.string().nullable().optional(),
  audio_path: z.string().nullable().optional(),
  clip_id: z.number().int().nullable().optional(),
  clip: ClipSchema.nullable().optional(),
  tag_id: z.number().int().optional(),
  tag: TagSchema,
  start_time: z.number(),
  end_time: z.number(),
  has_embedding: z.boolean(),
  is_active: z.boolean(),
  created_by_id: z.string().uuid().optional(),
  created_on: z.coerce.date(),
});

export type ReferenceSound = z.infer<typeof ReferenceSoundSchema>;

// Create from Xeno-Canto
export const ReferenceSoundFromXenoCantoSchema = z.object({
  xeno_canto_id: z.string().min(1, "Xeno-Canto ID is required"),
  tag_id: z.number().int(),
  name: z.string().min(1, "Name is required"),
  start_time: z.number().min(0).optional(),
  end_time: z.number().min(0).optional(),
});

export type ReferenceSoundFromXenoCanto = z.infer<
  typeof ReferenceSoundFromXenoCantoSchema
>;

// Create from dataset clip
export const ReferenceSoundFromClipSchema = z.object({
  clip_id: z.number().int().optional(),
  clip_uuid: z.string().uuid().optional(),
  tag_id: z.number().int(),
  name: z.string().min(1, "Name is required"),
  start_time: z.number().min(0).optional(),
  end_time: z.number().min(0).optional(),
}).refine(
  (data) => data.clip_id !== undefined || data.clip_uuid !== undefined,
  { message: "Either clip_id or clip_uuid is required" },
);

export type ReferenceSoundFromClip = z.infer<
  typeof ReferenceSoundFromClipSchema
>;

// Update reference sound
export const ReferenceSoundUpdateSchema = z.object({
  name: z.string().min(1).optional(),
  tag_id: z.number().int().optional(),
  start_time: z.number().min(0).optional(),
  end_time: z.number().min(0).optional(),
  notes: z.string().optional(),
  is_active: z.boolean().optional(),
});

export type ReferenceSoundUpdate = z.infer<typeof ReferenceSoundUpdateSchema>;

File: old/back/src/echoroo/schemas/reference_sounds.py
"""Schemas for Reference Sounds."""

from enum import Enum
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from echoroo.schemas.base import BaseSchema
from echoroo.schemas.clips import Clip
from echoroo.schemas.tags import Tag

__all__ = [
    "ReferenceSoundSource",
    "ReferenceSound",
    "ReferenceSoundCreate",
    "ReferenceSoundFromXenoCanto",
    "ReferenceSoundFromClip",
    "ReferenceSoundUpdate",
]


class ReferenceSoundSource(str, Enum):
    """Source type for a reference sound."""

    XENO_CANTO = "xeno_canto"
    """Reference sound imported from Xeno-Canto database."""

    CLIP = "clip"
    """Reference sound extracted from an existing clip in the dataset."""

    UPLOAD = "upload"
    """Reference sound uploaded directly by the user."""

    ANNOTATION = "annotation"
    """Reference sound derived from an existing annotation."""


class ReferenceSoundCreate(BaseModel):
    """Base schema for creating a reference sound."""

    name: str = Field(..., min_length=1, max_length=255)
    """Human-readable name for the reference sound."""

    tag_id: int = Field(..., description="Tag identifier for classification")
    """Tag that this reference sound represents."""

    start_time: float = Field(ge=0.0)
    """Start time of the sound segment in seconds."""

    end_time: float = Field(gt=0.0)
    """End time of the sound segment in seconds."""

    notes: str | None = Field(default=None, max_length=1000)
    """Optional notes about the reference sound."""

    @model_validator(mode="after")
    def validate_times(self):
        """Validate that start_time < end_time."""
        if self.start_time >= self.end_time:
            raise ValueError("start_time must be less than end_time")
        return self


class ReferenceSoundFromXenoCanto(ReferenceSoundCreate):
    """Schema for creating a reference sound from Xeno-Canto."""

    xeno_canto_id: str = Field(..., min_length=1, max_length=50)
    """Xeno-Canto recording identifier (e.g., 'XC123456')."""

    source: ReferenceSoundSource = Field(
        default=ReferenceSoundSource.XENO_CANTO,
        frozen=True,
    )
    """Source type (fixed to xeno_canto)."""


class ReferenceSoundFromClip(ReferenceSoundCreate):
    """Schema for creating a reference sound from an existing clip."""

    clip_id: int = Field(..., description="Clip identifier")
    """Clip to use as the basis for the reference sound."""

    source: ReferenceSoundSource = Field(
        default=ReferenceSoundSource.CLIP,
        frozen=True,
    )
    """Source type (fixed to clip)."""


class ReferenceSound(BaseSchema):
    """Schema for a reference sound returned to the user."""

    uuid: UUID
    """UUID of the reference sound."""

    id: int = Field(..., exclude=True)
    """Database ID of the reference sound."""

    name: str
    """Human-readable name for the reference sound."""

    ml_project_id: int = Field(..., exclude=True)
    """ML project that owns this reference sound."""

    ml_project_uuid: UUID
    """UUID of the owning ML project."""

    source: ReferenceSoundSource
    """Source type of the reference sound."""

    tag_id: int = Field(..., exclude=True)
    """Tag identifier for classification."""

    tag: Tag
    """Tag that this reference sound represents."""

    start_time: float
    """Start time of the sound segment in seconds."""

    end_time: float
    """End time of the sound segment in seconds."""

    duration: float = Field(default=0.0)
    """Duration of the reference sound in seconds."""

    xeno_canto_id: str | None = None
    """Xeno-Canto recording identifier if sourced from XC."""

    clip_id: int | None = Field(default=None, exclude=True)
    """Clip identifier if sourced from a clip."""

    clip: Clip | None = None
    """Hydrated clip information if sourced from a clip."""

    audio_path: str | None = None
    """Path to the cached audio file."""

    embedding_count: int = 0
    """Number of embeddings computed for this sound."""

    notes: str | None = None
    """Optional notes about the reference sound."""

    is_active: bool = True
    """Whether the reference sound is active for searches."""

    created_by_id: UUID
    """User who created the reference sound."""


class ReferenceSoundUpdate(BaseModel):
    """Schema for updating a reference sound."""

    name: str | None = Field(default=None, min_length=1, max_length=255)
    """Human-readable name for the reference sound."""

    tag_id: int | None = None
    """Tag identifier for classification."""

    start_time: float | None = Field(default=None, ge=0.0)
    """Start time of the sound segment in seconds."""

    end_time: float | None = Field(default=None, gt=0.0)
    """End time of the sound segment in seconds."""

    notes: str | None = Field(default=None, max_length=1000)
    """Optional notes about the reference sound."""

    is_active: bool | None = None
    """Whether the reference sound is active for searches."""

    @model_validator(mode="after")
    def validate_times(self):
        """Validate that start_time < end_time if both are provided."""
        if (
            self.start_time is not None
            and self.end_time is not None
            and self.start_time >= self.end_time
        ):
            raise ValueError("start_time must be less than end_time")
        return self
